net/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+pub type TransportAddress = Vec<u8>;
+pub type SerializedMessage = Vec<u8>;
+
+/// this closure type will get called when the send completes and the parameter will be the response message (or error)
+type SendResponseClosure = Box<FnMut(Result<SerializedMessage, Error>) -> Option<Error> + Send>;
+
+/// this closure type gets called when a new message arrives, you can respond with a Message or an error.
+type ReceiveClosure =
+    Box<FnMut(&SerializedAddress, &SerializedMessage) -> Result<SerializedMessage, Error> + Send>;
+
+pub trait Node {
+    fn get_address(&self) -> &SerializedAddress;
+    fn get_transport_address(&self) -> TransportAddress;
+}
+
+pub struct Handler {
+    pub handler: Option<ReceiveClosure>,
+}
+
+impl Handler {
+    fn handle(
+        &mut self,
+        from: &SerializedAddress,
+        message: &SerializedMessage,
+    ) -> Result<SerializedMessage, Error> {
+        match self.handler {
+            None => bail!("fish"),
+            Some(ref mut handler) => (handler)(from, message),
+        }
+    }
+}
+
+pub trait Transport {
+    /** initialize the transport
+     * this might be used for example in a TCP based transport to specify a listening port for
+     * for the transport.
+     */
+    fn initialize(&mut self, config: Option<String>) -> Result<(), Error>;
+
+    /** return a default config with best configuration guesses for the transport
+     */
+    fn get_default_config(&self) -> String;
+
+    /** register a peer as a node in the transport
+     */
+    fn new_node(&mut self, addr: SerializedAddress, handler: Option<Handler>) -> Result<(), Error>;
+
+    /** send a message to a node over the transport
+     * assumes that the sending address was registered locally with new_node
+     */
+    fn send(
+        &mut self,
+        from: &SerializedAddress,
+        to: &SerializedAddress,
+        msg: SerializedMessage,
+        callback: SendResponseClosure,
+    ) -> Result<(), Error>;
+
+    /** deliver a message to a given recipient
+     * assumes that recipient address was registered locally with new_node
+     * this function may be called by the routing layer to bridge across transports
+     */
+    fn deliver(
+        &mut self,
+        from: &SerializedAddress,
+        to: &SerializedAddress,
+        message: SerializedMessage,
+    ) -> Result<SerializedMessage, Error>;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{collections::HashMap, str};
+    //    use error::NetworkError;
+
+    pub struct SimpleNode {
+        hc_addr: SerializedAddress,
+        transport_addr: u32,
+    }
+
+    impl Node for SimpleNode {
+        fn get_address(&self) -> &SerializedAddress {
+            &self.hc_addr
+        }
+
+        fn get_transport_address(&self) -> TransportAddress {
+            format!("{}", self.transport_addr).into()
+        }
+    }
+
+    pub struct SimpleTransport {
+        config: String,
+        nodes: Vec<Arc<SimpleNode>>,
```

Comment:
Unless you don't want for those nodes to be mutable you should put them into a `Vec<Arc<Mutex<SimpleNode>>>`

Reply:
For a real transport implementation yes, but because this is just for the tests I left it simple.